### PR TITLE
Curve fitting prototype using gradient descent

### PR DESCRIPTION
**Pull request by zedfly**
Curve fitting prototype on the arduino using the Gradient-Descent method

### CODE REVIEW

File: prototype/gradient_descent/gradient_descent.ino [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+uint32_t degree = 1;
+uint32_t n = 1;
+gradient_real *x;
+gradient_real *y;
+
+void setup() {
+  Serial.begin(115200);
+}
+
+gradient_real cost(gradient_real p[]) {
```
Comment (nicholaslegrand102) [inline prototype/gradient_descent/gradient_descent.ino:19]: Please add comments describing the functions

File: prototype/gradient_descent/anomaly_detection.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+    line_x = np.linspace(x[0], x[-1], 100)
+    plt.plot(line_x, predict(coeff, line_x))
+    plt.show()
+    
+    port.close()
+
+if __name__ == "__main__":
+    detectAnomaly("data/Linear_20%_outliers.csv", 1, 300)
+    #detectAnomaly("data/Cubic_20%_outliers.csv", 3, 3000)
```
Comment (nicholaslegrand102) [inline prototype/gradient_descent/anomaly_detection.py:82]: missing newline

File: prototype/gradient_descent/anomaly_detection.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    # Send the points
+    for val in x:
+        port.write(struct.pack('<f', val))
+
+    for val in y:
+        port.write(struct.pack('<f', val))
+
+    port.write(struct.pack('<i', degree))
+
+    # Grab de coefficients of the regression from the arduino
```
Comment (nicholaslegrand102) [inline prototype/gradient_descent/anomaly_detection.py:22]: Spelling

File: prototype/gradient_descent/gradient_descent.ino [hunk 4]
```diff
@@ -0,0 +1,84 @@
+extern "C" {
```
Comment (nicholaslegrand102) [inline prototype/gradient_descent/gradient_descent.ino:1]: Extern is not needed for the import

File: prototype/utils/arduino_serial_port.h [hunk 5]
```diff
@@ -19,4 +19,8 @@ void writeElement(void* element,
                   const uint32_t sizeOfElement);
 
 
+<<<<<<< HEAD
```
Comment (nicholaslegrand102) [inline prototype/utils/arduino_serial_port.h:22]: Unfinished merge

File: prototype/gradient_descent/anomaly_detection.py (hunk 2 above)
Comment (zedfly) [reply prototype/gradient_descent/anomaly_detection.py]: done

**Review by bisson2000 — APPROVED** [review]
lgtm

File: prototype/gradient_descent_anomaly_detection/README.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
+# Gradient descent anomaly detection
+
+## Presentation
+
+The current prototype offers the necessary logic to perform simple anomaly detection on data points comming from a CSV file.
```
Comment (felix642) [inline prototype/gradient_descent_anomaly_detection/README.md:5]: comming

File: prototype/gradient_descent_anomaly_detection/README.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+To do so the arduino is sent the data points and will then compute the polynomial regression coefficients that best fits them.
+A python script will the use these coefficients to predict the next data point and check if it is close to the next data point from the CSV file.
```
Comment (felix642) [inline prototype/gradient_descent_anomaly_detection/README.md:7]: "will the use"

File: prototype/gradient_descent_anomaly_detection/anomaly_detection.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    # Let arduino have enough time to initialise serial connection
+    time.sleep(2)
+
+    n = minPoints + 1
+    while (n < len(x)):
+        coeff: list(float) = computeRegression(port, x[:n], y[:n], degree)
+
+        y_predict: float = predict(coeff, x[n])
+
+        # If the value is to far off the prediction. It is probably an anomaly
```
Comment (felix642) [inline prototype/gradient_descent_anomaly_detection/anomaly_detection.py:58]: to far off

File: Makefile [hunk 9]
```diff
@@ -44,6 +44,9 @@ gradient_descent: ./$(TEST_FOLDER)/test_gradient_descent.c ./src/gradient_descen
 monte_carlo: ./$(TEST_FOLDER)/test_monte_carlo.c ./src/monte_carlo.c ./src/linear_congruential_random_generator.c | build_folder
 	$(CC) $(CFLAGS) $^ -o $(BUILD_FOLDER)/test_$@.out $(LDLIBS)
 
+gradient_descent: ./$(TEST_FOLDER)/test_gradient_descent.c ./src/gradient_descent.c | build_folder
```
Comment (felix642) [inline Makefile:47]: Please remove this entry from the Makefile.

**Review by felix642 — CHANGES_REQUESTED** [review]

**Review by felix642 — APPROVED** [review]
LGTM!

**Pull request merged**